### ISSUE TITLE
KDBX 5 Standard

### ISSUE THREAD

**Issue by droidmonkey** [keepassxreboot/keepassxc#4317]
Placeholder for KeePass Database (KDBX) standard discussion. This post will be edited to include a list of features and upgrades we would like to bring to Dominik for consideration in the next KDBX format update.

**Must Haves:**
 - [ ] Standardize on Argon2id instead of Argon2d
 - [ ] Add Password Modification Datetime
 - [ ] Database/Entry Custom Data modification time
 - [ ] Built-in Entry Layout Specification (https://github.com/keepassxreboot/keepassxc/issues/863#issuecomment-680292058)
 - [ ] Built-in TOTP Specification (`otp` attribute)
 - [ ] Built-in Hardware Key Specification
 - [ ] Built-in Quick Unlock Specification
 - [ ] Built-in favorites/pinned items
 - [ ] Only support straight binary key files
 - [ ] Multi-valued core and custom fields
 - [ ] Proper storage format for and deduplication of BLOBs (SQLite can handle this)
 - [ ] Transactional edits/global history on database level, not entry level (e.g., to undo deletes or moves and to ensure overall consistency)

**Nice To Haves:**
 - [ ] Custom icon naming
 - [ ] Graph instead of tree structure (one item can belong to multiple groups, makes tagging somewhat redundant)

**Optional Extensions:**
  None so far

Long form discussions should be held here: #5819

**Comment by droidmonkey** [comment]
Looping in @keepassium @mmcguill @PhilippC
Original issue that spawned this conversation: https://github.com/keepassxreboot/keepassxc/issues/4293#issuecomment-583742873

**Comment by mmcguill** [comment]
Might be worth adding the idea of "Favourites" or "Pinned" items, items that appear upfront when user opens the database. It's a pretty popular/handy feature in Strongbox and it would be great to have it standardised across the KeePass world.

**Comment by georgesnow** [comment]
Curious do all the variant apps support tags? If so why not just standardize on a tag? Maybe “Pinned” (case insensitive). And document it that way?

That would mean those apps that want to support it can and it wouldn’t matter to any other platform (or native KeePass, MacPass,KyPass etc)

**Comment by georgesnow** [comment]
And if someone happens to add it as tag the first time just prompt them explaining the use of the “Pinned” as a tag

**Comment by droidmonkey** [comment]
@georgesnow that's exactly the reason to make it _part of the standard_ instead of just arbitrarily deciding on a tag and gaining application concurrence. Tags might not be the best choice anyway since they are user defined and can conflict.

**Comment by georgesnow** [comment]
I am certainly not disputing adding the standard, which is always a better solution for something like this. However, seeing as the native KeePass tends to be conservative in their approach to making such changes (and understandably so). 

The suggestion was more of interim solution (should have been clearer about that). That is just based on is there any reference for how long features take to get integrated in the KeePass standard? 

The interim solution could be easily revertible or convertible by an app at later time when the new standard is implementated. It provides the benefit to the users now, but doesn't force compliance and break anything in other KeePass ports. I am certainly not saying it's ideal as you said due to user conflict issues. Again was just a thought, but sounds like thinking is not to do an interim solution and just wait, which I get.

**Comment by droidmonkey** [comment]
Correct, I am trying to avoid implementing a multitude of "minor" interim solutions. That creates significant code bloat with exceptions and workarounds depending on which version of the standard you are on. The myriad of plugins made for KeePass over the years demonstrates why this is not a good thing. They all made their own unique standards.

**Comment by F3000** [comment]
Example from Enpass for tags : 

Here they are user defined and the developers from Enpass using a pre defined “Apple Watch” Tag to display all entries on the watch which get tagged with this . 

In my eyes “pinning” should be in the standard and tags only for special things a developer decide

**Comment by funnym0nk3y** [comment]
Is there a plan, when those features could be introduced in the standard and into XC?

**Comment by droidmonkey** [comment]
This would be a keepassxc 3.0.0 type move. Ideally we'd get Dominik on board with this as a well.

**Comment by funnym0nk3y** [comment]
Hmm, I guess 3.0.0 is still far away. Thanks anyway! Is somebody in contact with Dominik?

**Comment by droidmonkey** [comment]
Not yet, want to get our full proposal together first using actual rfc style.

**Comment by funnym0nk3y** [comment]
If you need help let me know!

**Comment by hokonch** [comment]
Related #398  #3558

KeePassDX, the alternative KeePass client for Android, recently added multiple URL formats that are incompatible with Keepass2Android. The app uses `WebDomain` and `WebDomain[counter]` instead of `KP2A_URL` and `KP2A_URL_<counter>`.
https://github.com/Kunzisoft/KeePassDX/issues/524#issuecomment-708251598

Given the popularity of Keepass2Android, it's reasonable for KeePassXC to support its vendor prefix at this time, but I'd like to see this feature standardized with a common prefix in the future.

**Comment by strongbox-mark** [comment]
That's a pity, are we too late to ask the KeePassDX team if they change/standardize? Is this gone to Production yet?

**Comment by droidmonkey** [comment]
I'd just bake a proper one into kdbx 5 at this point.

**Comment by J-Jamet** [comment]
Hello gentlemen, I read the discussion succinctly, I may have missed some information, please remind me if I am wrong.

> That's a pity, are we too late to ask the KeePassDX team if they change/standardize? Is this gone to Production yet?

Indeed, KeePassDX is already in production with the `WebDomain` label feature. For my part, I just did not use the KP2A nomenclature which clearly indicated that it came from KeePass2Android, prefix ("KP2A_") and preferred used a more generic one.

This functionality also includes fields with `ApplicationId[n]` label which allow opening applications directly from an entry. I just saw that KeePass2Android released its new version 1.08c-r1-nonet with this note "No longer storing package names of Android apps in the URL field". But I don't know what that means yet.

As I indicated in my comment, I am of course ready to harmonize things, which will be much better in the future, and I think custom label fields need to be more appropriate and generic.

**Comment by strongbox-mark** [comment]
Hi @J-Jamet - Yes, I hear you KP2A_ seems a little too specific, and I think KeePassXC uses kpxc_ for some of it's extensions (Audit Exclusions, etc).

Thanks for the extra info on your URLs and Applications - Sounds interesting and it may be possible to use these on iOS too...

Maybe we should just go with "kp_ext_*" or "kdbx5_" or something like that for features we think would be very useful additions?

**Comment by PhilippC** [comment]
please note that KP2A does not really care about any prefix or so. If it searches for an entry matching a URL, it will first look into the URL field but if nothing matches there, it will search all other fields. If KP2A creates a new field for an android packages name, this is now called AndroidApp[counter], but again, it's just a name which doesn't have a meaning. If a new standard would define new types like URL, AndroidApp etc. that might simplify the matching.

**Comment by J-Jamet** [comment]
Thank you @PhilippC for the details. 

So we can replace the `ApplicationId[n]` label by `AndroidApp[n]`, which is fine and will allow better compatibility. 
It does not prepare for IOS apps, maybe it is a good thing if we want to separate the links available for applications which would not exist on Android or IOS. But would not work with the same names of packages (It does not matter to me but maybe important with people who have more than one type of phone). To discuss.

For URLs, we will harmonize with a common prefix. But if the prefix is ​​neither important for KeePassDX, nor for KeePass2Android, the method of visualizing links from KeePassXC is maybe too restrictive and should apply to all custom fields if a link is present whatever the label.

**Comment by droidmonkey** [comment]
I would prefer to use the existing reserved keyword "Url" and add a common suffix _N] or just [N

**Comment by phoerious** [comment]
I like FIELDNAME[n]. The User-Interface could generally display fields following this pattern as multi-valued. In fact, for KDBX5 I would multi-valued fields into the specification.

**Comment by phoerious** [comment]
BTW I added the following suggestions to the list at the top, which been lingering in my head for some time:

- Multi-valued core and custom fields
- Proper storage format for and deduplication of BLOBs (SQLite can handle this)
- Graph instead of tree structure (one item can belong to multiple groups, makes tagging somewhat redundant)

**Comment by droidmonkey** [comment]
Great idea, conversion from single value to multi valued fields can be easy too when upgrading from kdbx4 to kdbx5.

**Comment by phoerious** [comment]
Or in SQLite by adding an order attribute. I am more and more in favour of using SQLite, especially considering storage of binary blobs.

**Comment by J-Jamet** [comment]
Multi-valued core and custom fields: I'm ok for this :)

OK, so to resume (for short-term changes, I plan to change the format for the next update):  
- URL: `URL[_N] `
- Android Application Id or Package Name: `AndroidApp[_N] `
- IOS App Id or Bundle Id: `iOSApp[_N] `
- For elements specific to an application or plugin, it's only in plugin data otherwise we can ask here if it is useful elsewhere.

In any case, `[_N]` means the first element has no suffix, the following ones start from `_1` (ie: `URL`, `URL_1`, `URL_2`).
_Acronyms can stay with capital letters_

Also, does everyone agree that the OTP format adopted by default is the [URI otpauth](https://github.com/google/google-authenticator/wiki/Key-Uri-Format)? 
Because it is a pain to maintain with all the plugins that exist on KeePass (as described above). 
Is it a problem to capitalize for you or for the current plugins (simply to harmonize)? -> label `OTP` and not `otp`.

@phoerious For SQlite, tell me if I'm wrong but are you thinking of using it as external storage tools, not directly in KeePass format? Is it a technical choice of KeePassXC to deal with BLOBs?

Graph instead of tree structure: Are you thinking of including elements in the XML format or is it just at the parser level?

Built-in Quick Unlock Specification: How do you see the quick unlock, because I realize it is a subject with multiple interpretations : 
- Is it leaving the database open and offering a visual unlocking?
- Does it close the database and temporarily store the master key to open with a temporary key?

**Comment by phoerious** [comment]
Is there a reason for prefixing the numbers with an underscore?

SQLite is a consideration for KDBX5 (or KDBS for that matter). It would replace the XML in order to provide a more scalable storage backend.

**Comment by J-Jamet** [comment]
> Is there a reason for prefixing the numbers with an underscore?

Yes, because @droidmonkey [prefers](https://github.com/keepassxreboot/keepassxc/issues/4317#issuecomment-728099811). :D Otherwise we don't care

> SQLite is a consideration for KDBX5 (or KDBS for that matter). It would replace the XML in order to provide a more scalable storage backend.

Okay, why not, it could be more efficient indeed.

**Comment by phoerious** [comment]
I believe that was meant to be Url_1, Url_2 etc., not Url[_1], Url[_2]. ;-)

**Comment by droidmonkey** [comment]
Prefix of _ to me just separates the attribute name from the sequencing. Makes it clearer what the name is

**Comment by J-Jamet** [comment]
@droidmonkey Tend to agree with this argument

@phoerious I don't understand what you mean. The brackets are just to explain the concept, I was suggesting `URL`, `URL_1`, `URL_2`. But do you want to have only the first letter in uppercase with acronyms and why?

**Comment by phoerious** [comment]
I guess it should be something that is not too unpleasantly technical when viewed with a non-conforming client, but explicit enough to not cause unwanted effects with existing attributes. For KDBX5, it's a non-issue, since multi-values fields need only one key.

**Comment by J-Jamet** [comment]
> For KDBX5, it's a non-issue, since multi-values fields need only one key.

I understand that KDBX5 or KDBS format is important and I agree with the new standardization.
But we also have to think about migration and backward compatibility.
Sounds harmless with uppercase and underscore, but not everyone will migrate overnight (KeePass app creators and users who might want to keep an old format for whatever reason).
I still have a lot of users on my side who are still on KDB so we have to harmonize as much as possible even old formats so that the migration goes smoothly.

**Comment by droidmonkey** [comment]
I think we can uniformly agree that kdb needs to be read only (or import only on our case). Kdbx 3 needs to be formally deprecated.

**Comment by J-Jamet** [comment]
> I think we can uniformly agree that kdb needs to be read only (or import only on our case). Kdbx 3 needs to be formally deprecated. I have no interest in letting people cling onto old, potentially less secure, data formats that they don't have any understanding of at all.

However KeePass 1.38 is still downloadable and still maintained by Dominik. Personally, I cannot pretend that the format is no longer in use even if I want. Of course, I push to go on the new versions, but either it is completely stopped by everyone or not.
If you have a list of security flaws in these older formats, I'm interested in reading it. We could use this argument to stop their uses altogether.

**Comment by droidmonkey** [comment]
Maintain multiple formats for no reason at all is a security flaw in and of itself. Opportunity for failures abound. Older versions of kdb and kdbx do not support modern KDF such as argon2 and do not have critical HMAC checks to verify data integrity. Not a security flaw, but certainly an issue.

**Comment by 0xpr03** [comment]
> KeePass 1.38 is still downloadable and still maintained by Dominik

From the Outside: I think it'd make sense to already involve them into this discussion. That way you could get some input what the constraints are for them or for example if there are any deprecation plans for 1.X. Otherwise you may end up doing all the work and getting a no at the end.

**Comment by J-Jamet** [comment]
> Older versions of kdb and kdbx do not support modern KDF such as argon2 and do not have critical HMAC checks to verify data integrity.

I agree on this point but like you said it's not a flaw. It surely deserves that we warn people to migrate, not that they are forced. 

> Just because someone makes something available doesn't mean you have to support it

Historically, KeePassDroid is compatible with this format. I could have completely abandoned it but I preferred to scale backwards compatibility by redoing the entire architecture of the project. It was clearly a choice! :) I don't recommend using this format (it was only a technical challenge). I am ready to remove it from KeePassDX in the future, that's not the problem.

> KeePassX is still available, we actively discourage its use since it hasn't been updated in 5 years

Same, but here KeePass 1.38 was released last year

> Things have to be retired, the world changes, I'm in the camp of "move or die".

Indeed I see that, I didn't think the open source world was like this. I'm not as categorical as you on the subject.

Now the fateful question : To return to the main point, does that mean you only want to use the new KDBX5 format when it's in place? RIP KDBX3/KDBX4?

@0xpr03 I completely agree, it's easiest to ask Dominik directly, I'll send him an email.

**Comment by 0xpr03** [comment]
> Indeed I see that, I didn't think the open source world was like this. I'm not as categorical as you on the subject.

(I think it applies even more to OS. People could pay you for supporting totally outdated stuff or maintain it by themselves. Otherwise you're doing the work of a corporation for 0€ which will end up in a burnout.)

**Comment by J-Jamet** [comment]
It was the sentence to convince! :heart: :dagger: :drop_of_blood:  
I just sent Dominik an email regarding the subject. It started with a small observation but an interesting debate.

**Comment by droidmonkey** [comment]
My plan is to hard deprecated kdbx3 with our 2.7 release (message to user, upgrade now please!) And move kdbx3 to import only when kdbx5 is complete.

**Comment by J-Jamet** [comment]
Okay, in this case I am also thinking of warning users for this purpose with major versions of KeePassDX to deprecate older versions of databases. This may take a little longer than for KeePassXC but will be done over time.

**Comment by strongbox-mark** [comment]
Great discussion. I think it would be great if we could get Dominik involved. It's his baby after all. I'm not sure how busy he is, or interested in pushing/extending the format? Please let us know how it goes @J-Jamet - Maybe he could join us here!

**Comment by J-Jamet** [comment]
@strongbox-mark I already sent him an email with the link to the issue, I'm not going to harass him. I'll tell you if I have an answer.

**Comment by phoerious** [comment]
I have seen quite a few people who insisted they still needed KDB oder KDBX3 or some shit and their reason has always been "the new format isn't compatible with my mobile app". After I pointed them to the correct apps, they usually switched right away.

I believe it is important to keep general support for the old formats around, but it can be import-only and you have to save in a new format. That is totally up to every single application, though. If you implement KDB[XS]5, nothing keeps you from fully supporting the old formats as well. I would recommend nudging users into the right direction, though. If you never deprecate things, users will forever keep using it. There will never be a point where you can discontinue something because nobody is using it anymore if you are not actively working towards that state.

**Comment by DReichl** [comment]
The KDBX 4 format does support more features and follows more best practices than the KDBX 3 and KDB formats, but I'm not aware of any real problems with the older formats (especially, I'm not aware of any practical cryptographic attack against them). There are apps that are still using the older formats and aren't maintained actively (which I don't really consider to be a problem as long as there aren't any security issues); deprecating the formats would unnecessarily break compatibility with them. Therefore, I'm not going to do that.

If you want to stop supporting the older formats (e.g. in order to reduce complexity), feel free to do that. For such a format, just show a message that it is unsupported by KeePassXC (calling it "deprecated" would be confusing).

I'll post my thoughts about the other ideas here soon. Thanks!

**Comment by droidmonkey** [comment]
Good point, we will refrain from using deprecated. We went a bit hard on the key file format, apologies

**Comment by J-Jamet** [comment]
Thank you @DReichl for these clarifications. As the old formats are not deprecated and do not represent security concerns. I will keep them until the implementation of new elements takes too much work for compatibility or migration.

**Comment by phoerious** [comment]
In the end it means the same thing and for the end user it for sure does. Deprecated doesn't mean bad, unsafe or anything, it simply means "unsupported in the future". Perhaps we should clarify that only KeePassXC (and some other apps) will stop supporting the old formats (or at least creating new files in that format).

**Comment by DReichl** [comment]
**Argon2id vs. Argon2d.**
Of the three Argon2 variants, Argon2d has the best resistance against GPU/ASIC attacks (which is important for database files), and the advantages of Argon2i are far less relevant in our case; see:
https://keepass.info/help/base/security.html#secdictprotect
Therefore, I think that Argon2d is the best choice in our case.

Is there a point that I'm overlooking?

**Comment by phoerious** [comment]
It's not an easy decision, actually. In theory, Argon2d is better for scenarios without a side-channel attack surface. This is probably the main scenario for an encrypted password file, but doesn't always have to be considering multi-user systems and the fact that most users have a web browser running all the time. I asked my colleague from the cryptography chair for an opinion and those were also his concerns, although in a case where these things actually matter, you are most likely dealing with state-sponsored adversaries and should take more precautions anyway (like not running your web browser while using KeePass and not using it on a shared system).

The main reason why we are considering Argon2i, however, is something as profane as tooling support. Argon2d has very little support in terms of libraries. The only reason we are using the original reference implementation is that libsodium (on which we already depend anyway), implements only Argon2i. We literally have to patch the argon2 library to rename symbols, so that both implementations do not clash at link time.

**Comment by phoerious** [comment]
Addendum: As I research this further, I see more tools implementing only Argon2i (PHP, for instance) and the general consensus being that Argon2d should NOT be used for password hashing. I guess this assumes online password hashing for authentication, though.

**Comment by J-Jamet** [comment]
> the general consensus being that Argon2d should NOT be used for password hashing. I guess this assumes online password hashing for authentication, though.

Do you have a link to the article that explains this consensus?
What about Argon2id? It would be a good thing to use the hybrid version, after that it remains to be seen if there are the adequate libraries.

**Comment by phoerious** [comment]
The consensus is what I perceived inspecting different sources. It is in no way authoritative. Argon2id in particular has even worse (or similarly bad) tooling support. I'm actually not quite sure why we have that on the list and not 2i. Otherwise it sounds like a jack of all trades.

**Comment by J-Jamet** [comment]
Because it's advocating, here's a document that explains different attack methods and why Argon2id is advocate by default.
I have to read it more precisely to give a more informed opinion, but it may be useful.
https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2/?include_text=1

**Comment by hokonch** [comment]
> libsodium (on which we already depend anyway), implements only Argon2i

According to the [libsodium documents](https://doc.libsodium.org/password_hashing),
> Libsodium supports Argon2i and Argon2id.

In addition to, the RFC draft says
>  Argon2id MUST be supported by any implementation of this document,
   whereas Argon2d and Argon2i MAY be supported.

**Comment by phoerious** [comment]
All right, then I misremembered and id is indeed what we had in mind. The current IETF RFC draft appears to be tending towards 2id as the go-to default. The current draft expires 12 March 2021. Perhaps we should wait until then and choose the recommended option.

**Comment by droidmonkey** [comment]
Sorry late to this, as @hokonch stated, Argon2id is the shall implement. All crypto tool providers are actively NOT implementing Argon2d. The only way to get it is to use the reference implementation library.

**Comment by DReichl** [comment]
On a server, side-channel attacks of course are relevant. If I'd implement a key derivation for a server, Argon2id also would be my first choice, and I'm not surprised that the RFC draft recommends this.

I think we agree that the resistance against GPU/ASIC attacks is important in our case (it actually was the primary motivation to add Argon2). By switching from Argon2d to Argon2id, we'd lose some of this resistance and gain some side-channel attack resistance instead. It isn't clear to me whether this trade is worthwhile. Janek, you mentioned attacks in multi-user systems and through web browsers. I'd like to learn more about such attacks. Do you have a link to an article/website where I can read more about this? Did anyone accomplish such an attack (extracting sensitive information from a different process on a client PC)? I searched the web, but only found attacks that aren't really interesting (e.g. detecting certain user activities by observing the CPU cache, but this is very far away from extracting sensitive information).

Shall we implement Argon2id as an option? Users who worry about side-channel attacks could then select Argon2id, whereas users who prefer the best resistance against GPU/ASIC attacks (like me) could select Argon2d. As the key derivation is extensible since KDBX 4, adding support for Argon2id is easy; the database format doesn't even need to be changed for this (only support for the new algorithm needs to be added). I'd be happy with this solution.

**Comment by droidmonkey** [comment]
Given that we have all implemented Argon2d already this isn't really a big concern of mine. I just wish libsodium implemented it so we didn't need a separate library. I'm ok with not adding more options and just keep using 2d.

**Comment by phoerious** [comment]
I am not really an expert on side-channel attacks, but from what I understand, there are at least conceptual PoCs for cache timing exploits via Javascript in a browser. The same attacks are of course also possible (and probably even easier) on multi-user systems such as shared hosting servers.

Argon2d becomes vulnerable if you can detect memory patterns, since its memory access is data-dependent and you can abort early if the pattern doesn't match. Argon2i prevents this by ensuring the access pattern in each iteration is the same. This data-independent array access method also makes Argon2i the slower of both variants, which is why it is recommended for password hashing (besides general resistance against timing attacks in online systems and multiuser environments). Argon2id sits in the middle and provides the best of both worlds, hence it is the recommended option unless you have a very good reason to use one of the other two.

Of course, Argon2i is totally capable of resisting GPU attacks. Bot variants make time-memory trade-offs extremely unfavourable. Overall, it is simply less resistant  and needs more rounds to provide the same level of security against this kind of attack. Hence, Argon2d would be used for things like PoW in blockchains where side-channel attacks are entirely irrelevant, but has little application elsewhere. In the end, however, a somewhat strong password and reasonable parameter choice should make your transformation strong enough with all three options so that you may as well start brute-forcing the AES key.

I cannot really give you good hints for where to look in the literature, but the Open Sesame paper by Wetzels (although a student paper?) is worth a read and I also found this Reddit thread to be quite interesting: https://www.reddit.com/r/crypto/comments/c8a299/cache_timing_attacks_against_argon2d/
I can ask my colleague if he has better literature recommendations (besides Biryukov et al. of course). 

In the end, I am all for going with the recommended standards and that is Argon2id. The benefits are better tested, well-analysed and proven as well as often faster implementations than for any niche variant. I also want to get rid of libargon2 (and the nasty binary patching), which isn't really meant for use in production the way other mainstream crypto libraries would be. Making both d and id an option wouldn't allow us to do this.

**Comment by DReichl** [comment]
Do you mean the following cache timing exploits via JavaScript in a browser?
https://www.cs.columbia.edu/2015/spy-in-the-sandbox/
Although interesting, this is one of those attacks that I don't really consider to be a threat for us; you can only detect certain user activities, not extract sensitive data (as mentioned explicitly in the article).

Thanks for the hint about the Open Sesame paper and the link to the Reddit thread. I've read them, but I'm still not seeing evidence that side-channel attacks are a real threat on normal client devices right now (in contrast to side-channel attacks against servers). There are ideas how such local attacks might work in theory, but I'm not aware of any successful attack in practice (probably due to the noise on client devices, etc.). This might change in the future though, thus I understand that some people worry about it.

All three Argon2 variants aim for GPU/ASIC resistance, but with respect to this, Argon2d is a bit better than Argon2id. By switching from Argon2d to Argon2id, we don't simply get some side-channel resistance for free; we lose some GPU/ASIC resistance. This is a good reason for using Argon2d, if you don't fear side-channel attacks.

The KDBX format is supported by different apps, on different platforms, using different cryptographic libraries, which may change over time. I understand that you want to simplify your code (reducing code complexity is a good thing), but I don't think that the availability of certain implementations of an algorithm should influence the decision whether to use the algorithm; the decision should only be based on whether the algorithm is suitable (from a design point of view, with security having the topmost priority, where security may have different facets like in our current case).

In light of these points, I think that it's best to support both Argon2d and Argon2id, and let the user choose. The documentation should clearly mention the advantages/disadvantages of the two variants and the current state of side-channel attacks on client devices, such that the user can make an informed decision.

I'll add support for Argon2id in KeePass now. This may take a while (I'm currently still working on something else that will take a few more days, and properly implementing Argon2id will also take some time), but I'm sure it'll make it into the next KeePass release. I'll post here when my implementation is ready.

Thanks!

**Comment by phoerious** [comment]
I think tooling support is a very strong argument and I definitely trust a well-tested implementation of a widely-deployed version of an algorithm a lot more than a niche implementation of a niche algorithm. That is the same reason I keep using AES, even if other algorithms may be better in theory and I am certainly not relying on some old reference implementation code.

> By switching from Argon2d to Argon2id, we don't simply get some side-channel resistance for free; we lose some GPU/ASIC resistance.

And that is not true per-se. You can compensate easily by sacrificing a bit of decryption time with a higher round setting. In practice I don't see any reason to trade the advantages of recommended and widely-deployed standards for rather hypothetical protection against targeted ASIC attacks. I believe if you have to be worried about that, your name is probably Edward Snowden and if you are in that position, then ramping up the rounds a bit isn't going to be a huge inconvenience.

> Do you mean the following cache timing exploits via JavaScript in a browser?
> https://www.cs.columbia.edu/2015/spy-in-the-sandbox/

Not necessarily exactly that, but something like that, yes. But I am also thinking about Spectre/Meltdown and similar low-level exploits. I don't see the risk of this becoming lower in the future and as more data and compute moves into the cloud, I can only imagine the attack surface to widen.

My main point, however, is that I don't like going against the standards without a VERY good reason. And "a bit" of ASIC resistance isn't going to cut it for me. I also don't see that any of the libraries are going to change their minds on this. If anything, the support for Argon2d will only get worse. I do agree with the libsodium maintainer Jonathan cited earlier that "Argon2d doesn't have any major benefits over Argon2id." and " It's not the one its authors recommend."

As I said, everybody is free to provide as many additional KDFs as they like. We also offer TwoFish as an additional block cipher (not even sure if anybody really uses that). But I think keeping Argon2d as the default is the wrong signal. The default should be something that is a) generally recommended for the purpose of password hashing and b) widely available and well-tested.

**Comment by phoerious** [comment]
@DReichl Since from your last comment, I can assume that Argon2id will certainly be an option (default or no default), I submitted a PR against our code base that adds support for it. Please see #5726.

The UUID I chose is `9e298b19-56db-4773-b23d-fc3ec6f0a1e6`.

**Comment by DReichl** [comment]
I've finished implementing Argon2id and have notified all KeePass port developers; you should have received an e-mail.

If anyone else wants to test it, here's the latest KeePass development snapshot:
https://keepass.info/filepool/KeePass_201203.zip
https://keepass.info/filepool/KeePass_201203_Source.zip

A database file for testing the compatibility can be found here:
https://keepass.info/filepool/KeePass_201203.kdbx
The master password is "123".

Updated documentation:
https://keepass.info/help/base/security.html#secdictprotect

**Comment by phoerious** [comment]
Cool. We'll merge ours then. I see you used the proposed UUID, so no changes required on our side.

**Comment by strongbox-mark** [comment]
Great! I'll aim to get this into Strongbox asap.

**Comment by phoerious** [comment]
Our's is merged into develop with a backport to 2.6.3 pending.

**Comment by J-Jamet** [comment]
Nice! Good work! will take a little more time for me but it's also scheduled for soon.

**Comment by PhilippC** [comment]
Just released a Keepass2Android build (1.08d-r0) in https://play.google.com/apps/testing/keepass2android.keepass2android with support for Argon2id

**Comment by DReichl** [comment]
Nice, thanks!

**Only support straight binary key files.**
We've discussed this already. The XML key file format has various advantages; they're documented here:
https://keepass.info/help/base/keys.html#keyfiles

KeePass 2.47 (which will probably be released in the beginning of January) will introduce version 2.0 of the XML key file format (I've notified all port developers, but I'm posting a part of it here again, in case anyone else wants to test it). The goal is to make it easier to create a backup of a key file on paper. Furthermore, there are changes in the user interface of KeePass that should encourage the correct use of key files. In detail, the following things have been added/changed:

* Enhanced XML key file format (added hash that allows to verify the integrity of the key; values are now encoded using hexadecimal characters in order to improve the readability).
* Added command 'Print Key File Backup' (in 'File' → 'Print').
* In the dialog that asks whether to print an emergency sheet, there's now an option 'Also print a key file backup' (on Windows), if the master key contains a key file.
* Added command 'Create Key File' (in 'Tools' → 'Advanced Tools'), which shows a dialog for creating a new key file (random key) or for recreating a key file from a printed backup.
* Changed default key file extension from 'key' to 'keyx'.
* Key file dialogs now only show '.keyx' and '.key' files by default.
* When clicking the key file 'Browse' button in the master key creation dialog and selecting a file that is not an XML key file, KeePass now asks whether to use the file as key file anyway.
* Improved key file loading/saving error messages.

Here's the latest KeePass development snapshot:
https://keepass.info/filepool/KeePass_201207.zip
https://keepass.info/filepool/KeePass_201207_Source.zip

Example package (containing a key file, a database file, a key file backup printout as PDF, and screenshots of the UI):
https://keepass.info/filepool/KeePass_201207_Ex.zip

**Comment by phoerious** [comment]
> Enhanced XML key file format (added hash that allows to verify the integrity of the key; values are now encoded using hexadecimal characters in order to improve the readability).

All right, I can get behind this. Hash checks for integrity are a nice thing to have. But I still find the other formats besides the normal hashed key file somewhat ridiculous.

> Key file dialogs now only show '.keyx' and '.key' files by default.

This will confuse many users. We do that by default when generating a new key file, but many users have key files that do not end with .key. In fact, from what I can see most people have no clue about what other key file formats KeePass actually understands and even less about what format it generates by default. When we started showing warnings about non-hash key files, people's confusion wasn't about why we stopped supporting those file formats. Instead, they were *convinced* they were using a "normal" or "arbitrary" file, because "that's how KeePass does it" and then they were confused why we would enforce a "special" key file format. Only after we told them to actually look at what kind of key file they had been using all that time, they noticed that they had an XML key file and not an "arbitrary" file. This thing is somewhat of a legend.

> When clicking the key file 'Browse' button in the master key creation dialog and selecting a file that is not an XML key file, KeePass now asks whether to use the file as key file anyway.

Okay, so that is about what we are doing with non-hash key files at the moment. To be clear, we are talking about a future file format version (KDBX5 or KDBS5 or what have you), so we don't have to be backwards compatible here. I would propose the following: For KDBX4, we keep supporting the other existing key files and for whatever comes next, we

- use the new XML format as the go-to default, since it has the obvious advantage of enabling error detection,
- hash everything that is not in the new XML format,
- drop support for the old XML, binary, and hex key files, since they are entirely silly.

**Comment by droidmonkey** [comment]
One important piece to implement if you are encouraging print out of the key file is a way to easily recreate it. The end user shouldn't have to retype XML and hash values. Recommend a way to just take the hex value key info and regenerate the rest of the file for the user.

**Comment by phoerious** [comment]
> (added hash that allows to verify the integrity of the key; values are now encoded using hexadecimal characters in order to improve the readability).

BTW why no base64? Would be the more compact standard encoding for such things. I see hex encoding hardly anywhere these days except for SHA hash values.

**Comment by Bill-cpp** [comment]
From the comments above, it sounds like you're considering changes to the default format and behavior of keyfiles currently created by KeePassXC. I'm glad you are, because the attached PDF is a proposal for verified keyfiles. My proposal is the basis for Dominik's Version 2 XML keyfiles. However, Dominik chose to implement only part of my proposal, and to implement it by extending his XML format, rather than by using a simple unstructured file format that would be more in line with KeePassXC keyfile philosophy. I've implemented verified keyfiles in a fork of KeePassXC. My implementation provides substantial usability advantages not available in the forthcoming KeePass 2.47.
[VerifiedKeyfiles.pdf](https://github.com/keepassxreboot/keepassxc/files/5679186/VerifiedKeyfiles.pdf)

**Comment by phoerious** [comment]
Do you have a link to your implementation? The proposal sounds intriguing, but I am very hesitant to implement *yet another* key file format when we are still trying to reduce the number of formats already supported. Choosing the correct parser for a file that doesn't tell you what its contents are by its extension becomes harder and more error prone the more formats we provide support for. For example, if you fill a file with random data and that data happens to be exactly 32 bytes long, it is handled differently than a file of 33 bytes. And a file where those 32 bytes are encoded in ASCII hex triggers yet another code path. It all becomes very convoluted and hard to explain with more formats. XML is still the easiest to get right if it's the only format that needs "special" treatment besides just hashing the contents, but a random base58 string could easily be mistaken for another format, making it impossible for anybody to open their databases if they happen to be using a 51-character ASCII file.

**Comment by Bill-cpp** [comment]
Yes, I am well aware of the dilemma of introducing a new keyfile format. I was never comfortable with the "anything goes" (hashed) format for keyfiles that Dominik introduced some time after his original 32-byte binary design. The hashed “design” (actually _absence_ of design) means that _any_ deliberately designed keyfile format, including all the other legacy designs, the XML design introduced later, and my verified keyfile design must be implemented and documented as _exceptions_ to the “catch-all” hashed design. This fact alone argues against using arbitrary files as keyfiles. But of course, it's too late now to change this.

The intent of my design is to support manual typing to create keyfiles from hardcopy  backup. Therefore, it would be easy for a user to type one or two characters too many or too few. Therefore, if an ASCII file has 51 plus or minus n characters (currently n = 3), my implementation assumes it to be a damaged verified keyfile, and an error message is emitted saying how many characters off it is. Of course, this makes the disambiguation problem even worse than what you stated.

I was surprised to see that Dominik implemented Version 2 XML keyfiles using a new file extension. As I see it, the .key extension for Version 2 XML would have worked just fine. But now that he's established a precedent for .keyx, verified keyfiles could also use this new extension. It would be very easy to distinguish the two formats, even with the + or - n. Or I suppose verified keyfiles could have their own extension: .keyv?

**Comment by phoerious** [comment]
>  This fact alone argues against using arbitrary files as keyfiles.

It could as well argue against using anything BUT hashed key files. There is also nothing that is stopping you from generating a hashed key file that is typable. Simply use an arbitrary ASCII file instead of unprintable bytes. The only thing that you definitely don't have is a checksum.

> But now that he's established a precedent for .keyx, verified keyfiles could also use this new extension

The extension is just a cosmetic change and don't dictate how it is interpreted. We only use it to decide which kind of key file to generate, but you are free to change the extension how ever you like after that.

**Comment by Bill-cpp** [comment]
> _It could as well argue against using anything BUT hashed key files._

I must not have communicated well. My point was that as soon as you say that any file at all can be used as a keyfile by hashing its contents, then any file that wants to be interpreted differently as a keyfile must be documented as an _exception_ to the rule.

> _The extension is just a cosmetic change and don't dictate how it is interpreted._ 

It doesn't have to be that way. It could work like, "... but if the file extension is .keyx, then ..." As I see it, not recognizing special file extensions but treating them all the same runs counter to how things work in practically every application in every operating system I've ever seen. Ignoring file extensions just adds another obstacle to introducing a new keyfile format. As I see it, treating .keyx as "special" might be an opportunity to start to break away from this limitation.

**Comment by Bill-cpp** [comment]
It sounds like we both agree that using random files as keyfiles is a dubious feature. If an adversary knows or guesses that you're using this kind of security by obscurity, it is practical for him to use a brute force search, even if the operating system doesn't assist him further by remembering recently opened files.

There are many reasons why people may not want to or be able to invest in a security key. Although naive users may misuse keyfiles, power users can greatly enhance security by using a keyfile. A keyfile allows a database to be encrypted with the full power of 256 bits of entropy. It is not practical to achieve anywhere near this amount of entropy with only a master password that is memorized, even for most diligent power user. In practice, many users will choose relatively weak master passwords for “convenience”, reducing entropy even further. With password databases commonly stored in the cloud, it may become practical for adversaries to “harvest” databases not protected with the full power that 256-bit encryption has the potential to provide. 

In my view, strengthening support for keyfiles would significantly improve KeePassXC, especially for power users. As of now, it is not practical to back up to paper the 128 byte binary keyfiles created by KeePassXC. Therefore, backup copies of KeePassXC keyfiles must now be backed up in digital form only, incurring additional security risk. Dominik's solution involves writing and maintaining additional GUI code. With verified keyfiles, no additional code is needed to support backup and restore from paper. An ordinary text editor can be used for both jobs.

**Comment by droidmonkey** [comment]
For the good of this issue, let's move additional Argon2id and key file discussions over to our new discussion forum: https://github.com/keepassxreboot/keepassxc/discussions/5819

**Comment by DReichl** [comment]
> Key file dialogs now only show '.keyx' and '.key' files by default.

Some users are using files that aren't suitable as key files (e.g. files of the operating system). Showing only '.keyx' and '.key' files in key file dialogs by default should discourage this. However, I agree that this may confuse some users, and I like your idea of using different filter defaults in the two master key dialogs, thanks! So, I've changed it to the following:

* When creating/changing the master key of a database, only '.keyx' and '.key' files are shown by default in the key file selection dialog. A confirmation/recommendation message is displayed if the selected file is not an XML key file (screenshot 'Screenshot_ArbFileQ.png' in the example package above).
* When entering the master key for opening a database, all files are shown by default. No confirmation/recommendation is displayed (because this would be annoying for users who want to use a non-XML format).

> To be clear, we are talking about a future file format version (KDBX5 or KDBS5 or what have you), so we don't have to be backwards compatible here. I would propose the following: ...

I don't see what the key file format has to do with the database file format. Nothing in a key file depends on data in a database file, and vice versa (except that you need the correct key for decrypting the database file of course).

Your proposal would make opening databases more complicated, because depending on the database file format, key files would need to be interpreted differently. A complexity reduction can only be achieved by completely removing the support for binary and hex key files. This would break backward compatibility though; I'm not going to do this in KeePass.

Each of the key file formats can store a 256-bit key, so there's no problem with any of the formats from a security point of view (right now). The code required for supporting the binary and hex formats is minimal, and I'm not planning to add more such formats; if longer keys are required in the future, these can be stored in XML key files.

So, KeePass will continue to support all of the key file formats. The XML format is recommended and used by default, but the others aren't deprecated. It would be great if you'd rephrase your dialog accordingly, in order to avoid confusing users.

> Recommend a way to just take the hex value key info and regenerate the rest of the file for the user.

The key file creation/recreation dialog of KeePass supports this (see the screenshot 'Screenshot_Recreate.png' in the example package above). The instructions on a key file backup printout recommend to use this dialog.

> The print out should just contain the key value, not the xml bits.

The dialog for recreating a key file is convenient, but it might be removed in the future (for whatever reason), or the user might switch to a different KeePass-compatible application in the future that doesn't have such a dialog. If the complete XML file is printed, the user can use a text editor to recreate the key file (which is mentioned as alternative in the instructions on the key file backup printout; it is a bit more work, but definitely doable). This would not be possible if the printout would contain only a part of the key file.

> BTW why no base64?

Readability. Base64 contains characters that can easily be confused (e.g. '1lI' and 'oO0'). This is not the case for hexadecimal characters. It's also not that much longer (not even twice as long). As recreating a key file is a rather rare task, typing a few more characters isn't really a problem (in contrast, confusing characters can be frustrating).

In general, Base64 is of course more suitable for encoding binary data in XML files (it's used within KDBX files, in KeePass configuration files, etc.). Using hexadecimal encoding (and the whitespace) in key files is an exception, because the user may create a backup on paper (which usually isn't the case for any of the other XML formats).

**Comment by phoerious** [comment]
> - When creating/changing the master key of a database, only '.keyx' and '.key' files are shown by default in the key file selection dialog. A confirmation/recommendation message is displayed if the selected file is not an XML key file (screenshot 'Screenshot_ArbFileQ.png' in the example package above).
> - When entering the master key for opening a database, all files are shown by default. No confirmation/recommendation is displayed (because this would be annoying for users who want to use a non-XML format).

That's pretty much how we do it.

> I don't see what the key file format has to do with the database file format. Nothing in a key file depends on data in a database file, and vice versa (except that you need the correct key for decrypting the database file of course).

Not per se. But for a new format, users would have to upgrade their databases anyway, so that would be a good point to break compatibility with existing key file formats also.

**Comment by DReichl** [comment]
**Built-in Entry Template Specification**

Supporting different entry types (i.e. presenting entry strings in different ways, in different arrangements) can have some advantages (e.g. certain entries may look nicer, field values can be restricted, etc.), but there also are various disadvantages. Entry types increase the amount of work/time required by the user for creating/maintaining entries. For example, KeePass intentionally does not support an e-mail address field, because this would require users to put additional time into customizing data transfer configurations; see https://keepass.info/help/base/faq_tech.html#email (these arguments also apply to 'Password' vs. 'PIN', etc.). Furthermore, when using custom fields extensively, entries cannot be displayed nicely in a list anymore (introducing group-specific views results in more problems), decreasing the usability. Limiting searches would be more difficult for users. And so on.

The direction in general is questionable. Entry types are usually requested for storing data that is outside the scope of a password manager (e.g. contacts / business cards, passport details, notes without accounts, bookmarks without accounts, tasks, ...). I prefer to optimize KeePass to be a good password manager, not a personal information manager (I'm not saying that PIMs are bad, it's just a different kind of application that I don't want KeePass to shift to).

In my opinion, the disadvantages outweigh the advantages. Therefore, I'm not going to implement this in KeePass.

If you want to implement it anyway, you can of course do that; the KDBX 4 format should already support everything you need. My suggestion would be to continue storing the user data in entry strings and type information either in `CustomData` items or also in entry strings.

There currently is only one KeePass plugin that provides such entry types already: KPEntryTemplates. If you want to support more field types or implement it differently than KPEntryTemplates, it would be great if you'd work together with the KPEntryTemplates developer on a common format for the type information, such that users who really want such functionality have one solution for both KeePass and KeePassXC.

On terminology. As you know, the term 'entry template' is used for an entry that acts as a template for the entry creation dialog (i.e. whose field values are filled in already when the entry creation dialog opens; documented on the https://keepass.info/help/v2/dbsettings.html page). The functionality above is different, and in order to avoid confusion, I'd suggest to use a different term for it, e.g. entry 'types', 'layouts', 'forms', or something like that.

PS: In the next 3 weeks, I'm finishing and publishing the next two KeePass releases (1.39 and 2.47). After that, I'll have a look at the remaining points.

**Comment by droidmonkey** [comment]
I prefer the name "Entry Layout" agree to not conflict with established terms.

**Comment by phoerious** [comment]
I added transactional edits on database level to the list. We talked about it several times, but forgot to add it. IMHO that's one of the most important features of all. KDBX in its current form cannot deliver that. If you delete an entry, it's gone. If you move it out of the database, it's gone. If you move it to another group, it's not gone, but there is no way to go back automatically.

**Comment by droidmonkey** [comment]
This would be a saved global history, correct? Almost like an audit log.

**Comment by phoerious** [comment]
Pretty much. That would also make an undo feature easier.

**Comment by strongbox-mark** [comment]
Just shipped Strongbox support for:

- Argon2id
- Key File v2
- {TOTP} & {TimeOTP} placeholders 
- KeePass's new TOTP custom fields 
   - TimeOtp-Secret-*
   - TimeOtp-Length
   - TimeOtp-Period
   - TimeOtp-Algorithm

👍

**Comment by AO-LocLab** [comment]
Under “Use a specific sequence for this association:”, the choices are in English, which will be confusing for international users who don’t read nor understand English. I’d recommend making such placeholders translatable.

**Comment by DReichl** [comment]
**Translatable Placeholders**

This has advantages and disadvantages, and I'm not sure yet whether I'll add this in KeePass. In any case, placeholders should be translated only on the user interface level, not on the file format level (i.e. within KDBX/XML files, English placeholders should be used). Other applications do it similarly (e.g. LibreOffice Calc uses translated function names in formulas in the UI, but English ones in ODS files).

**Database Custom Data Modification Time**

Good idea; I've added this now for KDBX 4.1. For details, please see:
https://keepass.info/help/kb/kdbx_4.1.html

The KDBX version number is 4.1, because right now any application that can read a KDBX 4 file can also read a KDBX 4.1 file (ignoring/losing the last modification times of custom data items). If a breaking feature/improvement is added before the next KeePass version, 4.1 will be changed to 5. So, this version number is preliminary.

**Entry Custom Data Modification Time**

Entries and groups are treated atomically: the entry/group has a last modification time, but not individual fields of the entry/group. This ensures that the information is consistent, i.e. that the fields fit together. There of course also are situations where treating fields individually would be useful, but overall I think that ensuring consistency is better. Therefore, no last modification times for custom data items in entries/groups have been added.

**Comment by DReichl** [comment]
**Global History / Audit Log**

> If you delete an entry, it's gone.

There are multiple stages for deleting entries:

1. When executing the default delete command on an entry in a regular group, the entry is moved into the recycle bin group.
2. When executing the default delete command on an entry in the recycle bin group, the entry is removed from the database and an item with the UUID of the entry is created in the 'DeletedObjects' list (which can later be used during a synchronization).
3. The 'DeletedObjects' list can be cleared (in KeePass: 'Tools' → 'Database Tools' → 'Database Maintenance').

When moving an entry into the recycle bin group, no entry information is lost yet. If the user changes his mind or if the deletion was accidental, he can move back the entry.

When permanently deleting an entry (stage 2), I think it's important that the entry data is really gone (except the UUID). Consider a user who wants to share parts of his database with someone else; he could make a copy of his database, delete all data that is not supposed to be shared, change the master key and then distribute the database. If there would be any sensitive information left somewhere in the database (e.g. in a global history), this would be a security issue. The UUID isn't really sensitive information, so the item in the 'DeletedObjects' list is no problem.

The items in the 'DeletedObjects' list are rather small, so for most users they don't cause real space problems. Anyway, there is a command for clearing this list, such that really all information of old entries is gone.

One could store the history entries of all entries in one common list (global history), but in my opinion it's better to store them directly in the entries. If they would be stored in one list, it would be easy to forget clearing history entries when permanently deleting an entry (resulting in security issues, as mentioned above).

KeePass has a function to find all recently modified entries ('Find' → 'Last Modified Entries'). However, one could implement a function that additionally inspects the history entries, in order to display a detailed audit log. By comparing the history entries of an entry, the log could even show which fields were changed. The user interface could allow selecting multiple changes and offer to roll them back at once. All the required data for such a function is already there.

> If you move it to another group, it's not gone, but there is no way to go back automatically.

Good idea; I've now added support for remembering previous parent groups:
https://keepass.info/help/kb/kdbx_4.1.html

**Comment by phoerious** [comment]
I disagree. Moving an entry around without any sort of historical knowledge about what happened to it is already close to "item is gone". There should be a track record of what happened to an entry, not just what happened to its attributes.

As for purging data before sharing a database, it's even easier with an audit log, because you can simply delete any history for all items older than X. If you set it to zero, it's the same as disabling history altogether (you can do that with individual entry histories as well), but you can just as easily keep the last few days and delete everything else.

It also makes the feature more discoverable, since it is easier to visualise to the user, because you have one big history and not 200 small ones. I bet (actually, I know for a fact) that many users don't even know about the history feature at the moment (and if they were to share an entry, they would unknowingly share the history it). Granted, you can improve the UX for single-entry history, but I believe it's much easier and overall more useful, understandable, and accessible if there is one global history.

**Comment by droidmonkey** [comment]
Building a report of the history (timeline) for existing entries is trivial. However there is a lot of information not covered by entry history like group moves and such that a global history log would be good for.

**Comment by phoerious** [comment]
And group renames or anything pertaining to groups for that matter. Keeping a record of previous groups per entry is a start, but it becomes problematic when a group is deleted.

**Comment by DReichl** [comment]
History entries are saved, because the old data may be useful in the future, also months/years later. The history entries of one entry don't have anything to do with the ones of a different entry. The database options for limiting entry histories reflect this: there are options for limiting an entry history based on the number of history entries and based on the estimated size, but not based on the time. There is a command for a time-based history deletion ('Tools' → 'Database Tools' → 'Database Maintenance'), but this is a manual operation that is only useful in special situations.

In contrast to that, you seem to have a different feature in mind. As far as I see it, you basically want to record all operations (including ones that don't change important data, e.g. moving or renaming groups), show them in a log and possibly allow to undo sequences of operations. Limiting such a log based on the time or the number of operations (entry-independent) makes sense here.

So, what we currently have is similar to a backup/storage software keeping older versions of files. Your idea is similar to an undo function instead, with the undo information being stored in the database.

Combining these two features into one is problematic. From a functionality point of view, the latter is a superset of the former (one could collect/create history entries for a specific entry from the log). However, there also are disadvantages. Recording all operations would require a huge amount of space; too much in my opinion. Maintaining such a log (e.g. erasing non-consecutive parts of it in certain situations like an export) would be difficult and error-prone (easily leading to security problems). It's questionable whether the functionality justifies the space and complexity problems; I don't think that there are many users who want to see/undo sequences of operations that happened a long time ago.

The ability to see/undo some of the most recent operations would certainly be useful though. If you want to add such a function, my suggestion would be to keep the log/undo information in the process memory only, until the database or the application is closed. This is how most other applications do it, and I don't see why our case should be an exception. It would possible to see/undo the most recent operations, and important information (in history entries) continues to be saved in the database file (usually for a longer time); I think this would be a good solution.

If you disagree and really want to store log/undo information in the database file, you can store it in custom data items (e.g. with time suffixes in the item names in order to ensure that KeePass merges them as intended during a synchronization, maybe encrypting values with entry-/group-dependent keys in order to avoid security problems). KeePass won't use these; I'll instead maybe add an undo function in the future that stores the undo information in the process memory.

**Comment by droidmonkey** [comment]
Memory based undo/redo is my only preference. Storage in the database would be for an audit log only, and is particularly useful in shared usage scenarios or keeshare based databases.

**Comment by DReichl** [comment]
**Proper storage format for and deduplication of BLOBs (SQLite can handle this)**

In KDBX 4, the contents of files attached to entries are stored in the inner header (compressed and encrypted, without Base64 encoding) and a deduplication is performed: each content is stored only once and can be referenced multiple times in the XML part.

All other data stored in a database file is significantly smaller. A deduplication wouldn't save much space here, especially considering that a part of the compression (GZip/Deflate, turned on by default) is to detect repetitions and store back-references instead (typically in a 32 KB sliding window only, which should be sufficient for "normal" databases though; history entries are stored near the actual entry).

So, I doubt that there is a different format that would require significantly less space. Especially, I haven't found any feature in the SQLite database file format that could result in considerably smaller files. Furthermore, one also needs to consider other factors (security, complexity, extensibility, ...). In our case, I don't think that a SQLite-based format would be more suitable than the current KDBX format.

Some users try to attach many large files (GBs) to entries. I consider this to be out of the scope of a password manager and recommend to use a specialized file encryption software (e.g. VeraCrypt) for this task instead (the password manager can store the password for the encrypted file container).

**Comment by droidmonkey** [comment]
There is a significant challenge with the attachments being in the inner header, when exporting XML you lose your attachments. We could keep them in the inner header, but when XML export you create attachment keys in the xml with the attachments as base64 encoded. Upon import of the xml they would be placed back into the inner header of the new database. I'm not sure if there is a significant difference between compressed attachments in binary form or compressed attachments in base64. The base64 would compressed again with the xml. It would certainly make reading the database a little easier.

**Comment by Fraetor** [comment]
Would increasing the size of the sliding window have a positive effect? The databases I tend to work with are around 500KB in size, though I suspect most of that is in favicons.

**Comment by phoerious** [comment]
The biggest issue with the current format is that BLOBs are indexed by order of occurrence. At least they should get a unique ID to ensure consistency . I have seen databases where the XML got out of sync with the header BLOB. Using a BLOB's hash as ID would be a neat trick that would ensure consistency of both the content and the referencing mechanism.

**Comment by DReichl** [comment]
The XML documents in a KDBX file and in an exported XML file differ slightly. In a KDBX 4 file, attachments are stored in the inner header. An exported XML file doesn't have such a header, thus attachments are stored in the 'Meta/Binaries' node of the XML document instead (compressed and Base64-encoded, with deduplication). So, storing attachments is supported in both cases.

When storing attachments in the inner header, the database file typically is a bit smaller, but the difference isn't that large (when using a good GZip/Deflate library with appropriate settings, Base64 data is compressed pretty well). Another advantage is that fewer data needs to be processed (no Base64 encoding, the XML serializer doesn't need to write all the Base64 data, and consequently fewer data is fed into the compression algorithm), resulting in a better performance. Furthermore, depending on the used XML serializer and the Base64 encoder, there may be implementation advantages (e.g. no large strings in the process memory, possibly allowing a better process memory encryption). In my opinion, the additional complexity required for storing attachments in the inner header isn't that huge. Therefore, I currently don't want to remove this feature.

> Would increasing the size of the sliding window have a positive effect?

RFC 1951 (which specifies the Deflate format) says: 'The representation used in the "deflate" format limits distances to 32K bytes ...'. So, increasing the size of the sliding window would be incompatible with standard Deflate implementations.

The Deflate64 variant uses a 64 KB sliding window. Unfortunately, it is incompatible with the Deflate format, and not all libraries support it (e.g. zlib does not). For testing, I compressed a 4.3 MB database XML file (containing attachments as well as many entries) using Deflate and Deflate64, resulting in 589 KB and 584 KB files respectively; not an impressive improvement.

I'd rather add support for a different compression algorithm, e.g. LZMA/LZMA2. The reason why I haven't added support for another algorithm yet is that I'm not sure whether this would really be worth it. The compression wouldn't be that much better (for example, the 4.3 MB database XML file mentioned above compresses to a 555 KB file using LZMA2) and a compression library would be required (LZMA(2) is rarely supported directly by frameworks), increasing the complexity. Currently, I'm happy with GZip/Deflate and prefer to wait; maybe another compression algorithm becomes more widely supported in the future.

> The databases I tend to work with are around 500KB in size, though I suspect most of that is in favicons.

Custom icons are stored in the PNG format (for various reasons), which uses Deflate compression internally. Trying to compress already compressed data usually saves only very few space, even when the second compression algorithm is better (w.r.t. size).

> I have seen databases where the XML got out of sync with the header BLOB.

This is a bug in the application that saved the database file, not a problem of the file format.

The ID of a content item is specific to one database file. While loading a database file, entries may reference contents by their IDs; when loading has finished, the IDs are meaningless and should not be remembered. While saving a database file, every content gets a new ID. When implementing it like this, IDs cannot get out of sync.

One could make the IDs persistent (e.g. UUID or hash of the content), but I don't see any real advantage. A disadvantage would be that the IDs would be longer.

**Comment by phoerious** [comment]
> This is a bug in the application that saved the database file, not a problem of the file format.

Yes and no. The file format makes errors like these possible without any means of catching them.

> One could make the IDs persistent (e.g. UUID or hash of the content), but I don't see any real advantage. A disadvantage would be that the IDs would be longer.

That's exactly what I proposed and you don't even need to save a full SHA256 hash.

**Comment by DReichl** [comment]
If the consecutive IDs would be replaced by UUIDs/hashes, some developers might think that it's a good idea to store the binary contents and the entries separately, which could easily result in synchronization problems though (this time in the sense that there are binary contents that aren't referenced by any entry anymore, or entries referencing binary contents that have accidently been deleted already).

UUIDs/hashes are unnecessary here. With consecutive IDs, it's clearer that they are specific to one database file.

**Built-in TOTP Specification**

This exists in the meanwhile; KeePass 2.47 has introduced support for a `{TIMEOTP}` placeholder:
https://keepass.info/help/base/placeholders.html#otp

**Comment by droidmonkey** [comment]
TIMEOTP? Why not TOTP. We already settled on TOTP placeholder.

**Comment by DReichl** [comment]
The name `{TIMEOTP}` (short for 'Time-based One-Time Password') is consistent with the name `{HMACOTP}` (short for 'HMAC-based One-Time Password'), which exists since 2010.

**Comment by 0xpr03** [comment]
Yeah but I've never heard of timeotp only TOTP as short version, see also [wikipedia](https://de.wikipedia.org/wiki/Time-based_One-time_Password_Algorithmus). And why use the longer version if the shorter is a standard name for users?

**Comment by Fraetor** [comment]
Existing implementations such as in KeePassXC use the `otp` attribute for TOTPs, with the specific method being specified in the content of the atribute.
```
otpauth://totp/AccountName?secret=AAAAAAAAAAAAAA&period=30&digits=6&issuer=ServiceName
```

**Comment by J-Jamet** [comment]
It's just that KeePass and KeePassXC have taken different development paths.
There is no need to manage others TOTPs in desktop applications because few users use more than one app on the same machine (and they are already in production). So no more need to argue at this level.
The problem is now to manage these different formats in secondary applications (those on other devices). 
I made the choice to manage already existing TOTP formats to satisfy users but there are too many of them (especially with the list of historical plugins) and it's horrible to define them consistently in a same shared base.

That's why this thread exists for new features. I don't want the problem to happen again, it would be a shame, and we all share this same philosophy of sharing. ;)

**Comment by DReichl** [comment]
When adding support for the `{TIMEOTP}` placeholder, I did consider using the `otpauth` format for the parameters, but decided against it, for the following reasons:

* An `otpauth` URI contains a label, consisting of an account name and possibly an issuer name. For the purpose that the format has been designed (namely to transfer OTP generator information through a QR code), this is perfectly reasonable. However, in the case of KeePass, these names would duplicate the user name and the title of the owning entry (with possibility of desynchronization, etc.). One could use an empty label, but this would make the `otpauth` URI unsuitable for a transfer, which could confuse users. I think it's better to separate the storage format and `otpauth` URIs. For generating/importing proper `otpauth` URIs (suitable for transfers), commands can be provided.
* The `otpauth` format supports Base32 secrets only. This restriction is not present in RFC 6238, and thus I'd like to support additional encodings (UTF-8, hexadecimal, Base64).
* Entering/managing information in separate fields is clearer and easier for users.
* It makes sense to make specifying parameters for the `{TIMEOTP}` placeholder analogous to specifying parameters for the `{HMACOTP}` placeholder.

`{TOTP}` doesn't collide with `{TIMEOTP}`. So, if you want to continue supporting your solution, I don't see a problem with that.

**Comment by DReichl** [comment]
**Custom icon naming**

Good idea; I've added this now (and last modification/deletion times, for synchronizations):
https://keepass.info/help/kb/kdbx_4.1.html

**Comment by phoerious** [comment]
That's a quite useful. Although overall, I'd be happy if we could somehow sketch out what a KDBX 4.1 and/or 5 should look like together. I don't want to add too much bureaucracy to the process, but Dominik just picking or rejecting things and informing us about it, seems kinda one-way (no offence in any way, I appreciate the communication).

**Comment by 0xpr03** [comment]
>    [..] names would duplicate the user name and the title of the owning entry (with possibility of desynchronization, etc.). [..]  `otpauth` format supports Base32 secrets only. This restriction is not present in RFC 6238, and thus I'd like to support additional encodings (UTF-8, hexadecimal, Base64).

That sounds like a good reason to maybe migrate keepassxc towards the timeotp format ?

**Comment by Fraetor** [comment]
If it was made a part of the format standard that sounds like a good idea. The point of standards is to standardise on them. The transition path does need to be carefully considered however, and it may be worth having both the `otp` and `timeotp` attributes for a time while other clients move away from `otp` (such as mobile clients).

**Comment by mstarke** [comment]
> TIMEOTP? Why not TOTP. We already settled on TOTP placeholder. Either way we won't be supporting that specification, the key-uri syntax is far easier to deal with and condenses everything into one attribute (otp).

I know it's a hassle to implement a multitude of different formats but that's the deal when using a shared format. Might I urge you to reconsider not supporting another format? This would lead to KeePassXC special behaviour that other clients have to support as well. Regarding the ```otpauth``` field. You seem to have chosen to use a non-standard encoder field to add support for steam. Or is this a supported way of expressing different TOTP flavours?

**Comment by droidmonkey** [comment]
To be fair, keepass2 is late to the game with totp.

**Comment by droidmonkey** [comment]
@0xpr03 neither of those points are valid. The username and title can very easily be placeholders that are dynamically filled when polled. Base32, Hex, and whatever else are encoding schemes. You can convert to any of them at any time. Standardizing on Base32 actually makes everything easier because you don't need to guess what encoding scheme is being used and have extra bloat code to deal with that. So no, the proposed (or forced) keepass2 standard is terrible in my opinion.

**Comment by DReichl** [comment]
> Dominik just picking or rejecting things and informing us about it, seems kinda one-way

My posts weren't intended to only inform you. I'm not mentioning it in every post/e-mail, but as stated in red on the top of the KDBX 4.1 page and in my initial e-mail about KDBX 4.1 to everyone, nothing is final yet and all feedback/discussion is welcome.

In the cases where I added something, your suggestions were sufficiently clear to me and the implementation was straightforward (for example, the last modification time elements should be called 'LastModificationTime' for consistency with already existing 'LastModificationTime' elements, they should store the time in the same format, etc.). In the cases where I didn't want to add something, I tried to explain why. If you want something to be implemented differently or believe that I didn't think about something, then please don't hesitate to write that. All of the things mentioned on the KDBX 4.1 page are open for discussion, and until it's final, I have no problem at all with changing them.

> The username and title can very easily be placeholders that are dynamically filled when polled.

As the user name and the title would need to be URI-encoded, a minimal `otpauth` URI (assuming default values for the length, period and algorithm) would look like this:

`otpauth://totp/{T-CONV:/{TITLE}/Uri/}:{T-CONV:/{USERNAME}/Uri/}?secret=...`

or

`otpauth://totp/{T-CONV:/{USERNAME}/Uri/}?secret=...&issuer={T-CONV:/{TITLE}/Uri/}`

This would indeed avoid a desynchronization of the title and the user name.

I prefer the solution of the `{TIMEOTP}` placeholder, but like I wrote previously, there should be no problem if you want to continue supporting your solution.

> Standardizing on Base32 actually makes everything easier because you don't need to guess what encoding scheme is being used and have extra bloat code to deal with that.

There is no need to guess the encoding of a `{TIMEOTP}` or `{HMACOTP}` secret. The name of the string storing the secret ends with the encoding ('-Hex', '-Base32', '-Base64', no suffix for UTF-8), i.e. there is no ambiguity.

One of the advantages of making the parameters for `{TIMEOTP}` analogous to the ones for `{HMACOTP}` is that code can be reused. KeePass uses the same method for loading the secret (with the first part of the string name as parameter), i.e. no extra code is necessary.

**Comment by Fraetor** [comment]
Thanks for the clarification Dominik, the way the page was in the "help" section of the KeePass website, and its inclusion in 2.48 made it seem a bit more final. Perhaps the Introduction section could be changed as it currently reads (to me at least) as if KDBX 4.1 is the default, where it is really that a KDBX 4.1 database will not be downgraded if opened.

Maximising code reused is a sensible goal, but there are lots of implementations which have different code to be reused. Do we know what is the most commonly adopted method? Is `{HMACOTP}` widely implemented? If it is, then the code for that should be reusable (or rewritable into reusability), while if `otpauth` has wider implementation across KDBX compatible software, it may make sense to minimise disruption to their codebases. I'm not personally sure of how pervasive either these are in the wider ecosystem however.

From a user perspective `otpauth` should probably be supported as an import/export format, as it is the most common way secrets are given out (typically in QR codes), but that is out of scope for the database format, and can be easily reconstructed from individual fields if required.

With niche things like Steam, are the details stored in the same `otp` field as regular TOTPs?

**Comment by droidmonkey** [comment]
It is my understanding that HMACOTP has no support in any alt-client, mobile or desktop. KeePassXC obviously does not support it. I have no problem supporting the TIMEOTP placeholder that actually fills the calculated TOTP value,  my problem is with the extra attributes defining the parameters. We settled on key-uri for the exact reason of QR code standardization and it also supports HOTP if we decide to implement that. I don't see us moving off that as our default when creating new OTP settings for an entry. So by that virtue, we won't be compatible with KeePass unless you implement the otp parameter with key-uri. That is why I am worked up about this.

**Comment by J-Jamet** [comment]
From the beginning of unique codes implementation in KeePassDX, the app correctly manages the HmacOTP parameters of KeePass2, but not the associated placeholders {HMACOTP} / {TIMEOTP} (not useful because there is an internal display that standardizes this views elements, it needs to be improved). Also, I already naturally managed the otpauth://hotp links as well to be consistent.
I let you debate this part because you have different opinions but just to say these settings were already managed before.

**Comment by phoerious** [comment]
> We settled on key-uri for the exact reason of QR code standardization and it also supports HOTP if we decide to implement that.

That. Of course, we *could* wrap some magic around it to let the user copy it as a URL for export, but why should we? The raw field value should be usable as is and a {TIMEOTP} value makes no sense outside the application unless I manually reen code it and assemble it to a URL. Also having it as a URL allows us to parse the value using standard off-the-shelf tools. 

Regarding "we are free to implement it differently", sure, I know that. And we may. But everybody implementing things differently, because we cannot agree on anything is what drives open source apart. Having ten incompatible implementations in the end is the worse outcome for everybody, developers and users alike; even if each implementation has its own unique benefit. "Linux on the desktop" is a long history of exactly that kind of failure over and over again.

**Comment by phoerious** [comment]
I'm moving this thread to discussions.
